components/script/dom/urlpattern.rs

```diff
@@ -0,0 +1,800 @@
+            Options::default(),
+            &mut out.port,
+        );
+
+        // FIXME: Step 14: respect ignore case option from here on out
+        let _ = options;
+
+        // FIXME: Steps 15-16: Compile path pattern correctly
+        Component::compile(
+            processed_init.pathname.as_deref().unwrap_or("*"),
+            Options::PATHNAME,
+            &mut out.pathname,
+        );
+
+        // Step 17. Set urlPattern’s search component to the result of compiling a component given
+        // processedInit["search"], canonicalize a search, and compileOptions.
+        Component::compile(
+            processed_init.search.as_deref().unwrap_or("*"),
+            Options::default(),
+            &mut out.search,
+        );
+
+        // Step 18. Set urlPattern’s hash component to the result of compiling a component given
+        // processedInit["hash"], canonicalize a hash, and compileOptions.
+        Component::compile(
+            processed_init.hash.as_deref().unwrap_or("*"),
+            Options::default(),
+            &mut out.hash,
+        );
+
+        // Step 19. Return urlPattern.
+        // NOTE: not necessary since we use an out parameter
+        Ok(())
+    }
+
+    /// <https://urlpattern.spec.whatwg.org/#url-pattern-has-regexp-groups>
+    fn has_regexp_groups(&self) -> bool {
+        self.protocol.has_regexp_groups ||
+            self.username.has_regexp_groups ||
+            self.password.has_regexp_groups ||
+            self.hostname.has_regexp_groups ||
+            self.port.has_regexp_groups ||
+            self.pathname.has_regexp_groups ||
+            self.search.has_regexp_groups ||
+            self.hash.has_regexp_groups
+    }
+}
+
+impl Component {
+    /// <https://urlpattern.spec.whatwg.org/#compile-a-component>
+    fn compile(input: &str, options: Options, out: &mut Self) {
+        // Step 1. Let part list be the result of running parse a pattern string given input, options,
+        // and encoding callback.
+        let part_list = parse_a_pattern_string(input, options);
+
+        // Step 2. Let (regular expression string, name list) be the result of running generate a regular expression and
+        // name list given part list and options.
+        let (regular_expression_string, name_list) =
+            generate_a_regular_expression_and_name_list(&part_list, options);
+
+        // Step 3. Let flags be an empty string.
+        // Step 4. If options’s ignore case is true then set flags to "vi".
+        let _flags = if options.ignore_case {
+            "vi"
+        }
+        // Step 5. Otherwise set flags to "v"
+        else {
+            "v"
+        };
+
+        // Step 6. Let regular expression be RegExpCreate(regular expression string, flags).
+        // If this throws an exception, catch it, and throw a TypeError.
+        let cx = GlobalScope::get_cx();
+        rooted!(in(*cx) let mut regular_expression: *mut JSObject = ptr::null_mut());
+        new_js_regex(
```

Comment:
This will need to be updated to make the flags configurable.

Reply:
Right, I forgot about the flags... 
Fixed.